[tf.data] Switch to TF combinations instead of test_util

kvignesh1420 opened the pull request:
This PR switches the tests present in `tensorflow/python/data/util/` to use TF combinations instead of the `test_util` decorator to express the combination of graph/eager modes with TF1/TF2 environments for execution.

Addresses point 1 in https://github.com/tensorflow/tensorflow/pull/46761#issuecomment-770059963
cc: @jsimsa

TEST RUNS (UPDATED):
```
INFO: Build completed successfully, 304 total actions
//tensorflow/python/data/util:random_seed_test                           PASSED in 2.2s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:sparse_test                                PASSED in 1.5s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:convert_test                               PASSED in 1.4s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:traverse_test                              PASSED in 1.4s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:nest_test                                  PASSED in 1.4s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:options_test                               PASSED in 1.3s

INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/util:structure_test                             PASSED in 1.7s

```

kvignesh1420 committed:
use combinations in convert_test

kvignesh1420 committed:
use combinations in random_seed_test

kvignesh1420 committed:
use combinations in sparse_test

kvignesh1420 committed:
use combinations in structure_test

kvignesh1420 committed:
use combinations in traverse_test

kvignesh1420 committed:
removed test_util based parent class

kvignesh1420 committed:
add test_base dependency to BUILD file

kvignesh1420 committed:
modify parent classes for testing

kvignesh1420 committed:
reverse decorator order to address failing tests

kvignesh1420 committed:
switch to combinations from parameterized

kvignesh1420 committed:
lint fixes

jsimsa commented on tensorflow/python/data/util/convert_test.py:
```diff
@@ -18,16 +18,19 @@
 
 
-class ConvertTest(test.TestCase):
+class ConvertTest(test_base.DatasetTestBase, parameterized.TestCase):
```
All tests should use the `combinations.generate` decorator to identify which combinations of (TF 1 vs TF 2) x (graph vs. eager) to test. In other words, you should add the decorator even to methods that were not previously decorated with `test_util`.

kvignesh1420 replied:
addressed in the commit.

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -18,18 +18,21 @@
+class RandomSeedTest(test_base.DatasetTestBase, parameterized.TestCase):
 
-  @test_util.run_in_graph_and_eager_modes
+  @combinations.generate(test_base.default_test_combinations())
```
The `test_cases` variable below should be moved to the combinations definition instead. In other words, instead of having a single test that runs a for all for different parameters combinations, we should be running one test for each combination (and that's what the `combinations.generator` decorator is used for).

jsimsa replied:
This comment applies to all other tests that use the `test_cases` pattern.

kvignesh1420 replied:
addressed in the commit.

jsimsa commented on tensorflow/python/data/util/structure_test.py:
```diff
@@ -50,8 +50,7 @@
+class StructureTest(test_base.DatasetTestBase, parameterized.TestCase):
 
   # pylint: disable=g-long-lambda,protected-access
   @parameterized.named_parameters(
```
This should be replace with using `combinations.generate`. Here is an example of how you can create a product of test combinations: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/shuffle_test.py#L224-L227

kvignesh1420 replied:
addressed in the commit.

jsimsa commented on tensorflow/python/data/util/structure_test.py:
```diff
@@ -73,14 +72,14 @@ class StructureTest(test_base.DatasetTestBase, parameterized.TestCase,
       ("Nested_2", lambda: {
-          "a":
-              constant_op.constant(37.0),
+          "a": constant_op.constant(37.0),
```
You should avoid reformatting code that you are not changing.

kvignesh1420 replied:
addressed in the commit.

jsimsa reviewed (changes requested):
Thank you for the change. Did a pass and left some comments.

kvignesh1420 committed:
add combinations to all tests

kvignesh1420 committed:
refactor test_cases loop with combinations

kvignesh1420 committed:
use NamedObject in sparse_test

kvignesh1420 committed:
use lambdas while parameterizing

kvignesh1420 commented:
@jsimsa thanks for the info. I have refactored the tests and addressed a few sanity issues as well. Please let me know.

kvignesh1420 committed:
add info on lambdas in parameterized tests

jsimsa commented on tensorflow/python/data/util/convert_test.py:
```diff
@@ -18,35 +18,43 @@
+from absl.testing import parameterized
+
 from tensorflow.python.data.util import convert
+from tensorflow.python.data.kernel_tests import test_base
```
imports should be lexicographically sorted, so `tensorflow.python.data.kernel_tests` should come before `tensorflow.python.data.util` ... this comment applies throughout this PR

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -18,43 +18,98 @@
+  @combinations.generate(
+      combinations.times(
+          test_base.default_test_combinations(),
+          combinations.combine(test_case_fn=[
```
move the definition of the test cases into a helper function to avoid the excessive indentation ... see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/optimization/filter_fusion_test.py for an example

jsimsa replied:
This comment applies throughout this PR.

kvignesh1420 replied:
addressed in the commit

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -64,19 +119,14 @@ def testRandomSeed(self):
 
+    check(tinput=tinput, toutput=toutput)
+
     if not context.executing_eagerly():
```
Checks like these should be avoided when TF combinations are used. Instead, you should create separate test combinations for eager mode and graph mode.

kvignesh1420 replied:
I have removed the redundant test and all the modes are covered in `default_test_combinations`

jsimsa reviewed (changes requested):
Thank you very much for the new changes. I left a couple of comments, two of which applies throughout the PR (lexicographical ordering of importants and using a separate method to create test combinations, which allows you to use meaningful argument names).

kvignesh1420 committed:
reorder imports

kvignesh1420 committed:
refactor parameterized test cases

kvignesh1420 committed:
remove redundant test in random_seed_test

kvignesh1420 commented:
@jsimsa thanks for pointing me to the references. I have addressed the review comments. Please let me know how it looks.

In the functions which return `test combinations`, I have followed an approach in which all the parameters are `combinations.NamedObject`(s). This approach facilitated in resolving duplicate test name issues and keeps the parameterized values consistent with the arguments that the test cases need.

kvignesh1420 committed:
lint fixes

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -18,64 +18,120 @@
+  )
+  def testRandomSeed(self, input_fn, output_fn):
 
-    if not context.executing_eagerly():
```
Where is the equivalent of L67 through L78 being tested now?

kvignesh1420 replied:
covered it in a new test now. Somehow missed adding this.

jsimsa replied:
the newly created test has no assertion, so it is not equivalent to the remove code

kvignesh1420 replied:
@jsimsa the assert has been moved to the common `_evaluate()` method for both the tests.

jsimsa replied:
I see. The name of the `_evaluate` method mislead me. Consider renaming the method to `_checkEqual` or `_assertEqual`

kvignesh1420 replied:
addressed in the commit. Used `_checkEqual` method instead of `_evaluate`

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -18,64 +18,120 @@
-        msg = 'test_case = {0}, got {1}, want {2}'.format(
-            1, (g_seed, op_seed), toutput)
-        self.assertEqual((g_seed, op_seed), toutput, msg=msg)
+    tinput, toutput = input_fn._obj(), output_fn._obj() # pylint: disable=protected-access
```
Why is the private `_obj` API needed? The test combinations in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/optimization/map_vectorization_test.py pass functions as parameters as well and do not need that API.

jsimsa replied:
This applies throughout this PR.

kvignesh1420 replied:
I have removed the access to `_obj` wherever functions are being used within NamedObjects as calling NamedObject instances will internally call the function itself. However, in some cases, where `types` or `classes` are being layered with NamedObject's, the call to `_obj` is required to get hold of the actual data.

For instance, in `structure_test.py`:

```python
def testConvertLegacyStructure(self, output_types, output_shapes,
                                 output_classes, expected_structure):
    output_types = output_types._obj
    output_shapes = output_shapes._obj
    output_classes = output_classes._obj
```

Changes have been made accordingly.

jsimsa commented on tensorflow/python/data/util/sparse_test.py:
```diff
@@ -18,58 +18,405 @@
-            "classes": (((sparse_tensor.SparseTensor))),
-            "expected": True
-        },
+# NOTE(vikoth18): Arguments of parameterized tests are lifted into lambdas to make
```
What happens when you do not use lambdas?

kvignesh1420 replied:
I ran the tests without lambdas in `random_seed_test` (a similar situation to what we have in sparse_test).

If we do not use lambdas, then for the test combination of `(mode='graph', tf_api_version=2)`, the tests throw :
```console
File "/root/.cache/bazel/_bazel_root/f04f3979d0f8981d3682673d5f0c4874/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/util/random_seed_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 1513, in convert_to_tensor
    raise RuntimeError("Attempting to capture an EagerTensor without "
RuntimeError: Attempting to capture an EagerTensor without building a function.
```

Reference: https://github.com/tensorflow/tensorflow/blob/f6957507f30597fd458028de1470010312b979f8/tensorflow/python/framework/ops.py#L1495-L1515

Due to this reason, The tests in `structure_test.py` were already using lambdas. Thus instead of splitting existing tests based on different execution modes, using lambdas can remove that redundancy as it works with all test_base combinations.

jsimsa replied:
thanks for the detailed response

jsimsa commented on tensorflow/python/data/util/random_seed_test.py:
```diff
@@ -18,64 +18,120 @@
+      # and output from get_seed:
+      # (output_graph_seed, output_op_seed)
+      (
+          "CASE_0",
```
replace "CASE" with "test_case" throughout this PR

kvignesh1420 replied:
Replaced "CASE" with "TestCase" to be consistent with the style of names of NamedObjects in other tests. Let me know if it's ok.

kvignesh1420 committed:
call NamedObject functions directly

kvignesh1420 committed:
replace CASE with TestCase

kvignesh1420 committed:
add incremental random seed test

kvignesh1420 committed:
refactor random_seed_test

jsimsa commented on tensorflow/python/data/util/sparse_test.py:
```diff
@@ -18,58 +18,405 @@
+        classes_fn=combinations.NamedObject(
+            "classes_fn.{}".format(name), classes_fn
+        ),
+        expected=combinations.NamedObject(
```
I don't think `expected` needs to be wrapped in a NamedObject here (what benefit does it provide?). Not making it a `NamedObject` avoids the need for `_obj` later.

kvignesh1420 replied:
addressed in the commit.

jsimsa commented on tensorflow/python/data/util/structure_test.py:
```diff
@@ -49,41 +50,716 @@
+        value_fn=combinations.NamedObject(
+            "value_fn.{}".format(name), value_fn
+        ),
+        expected_structure=combinations.NamedObject(
+            "expected_structure.{}".format(name), expected_structure
+        ),
+        expected_types=combinations.NamedObject(
+            "expected_types.{}".format(name), expected_types
+        ),
+        expected_shapes=combinations.NamedObject(
+            "expected_shapes.{}".format(name), expected_shapes
+        )
```
you should either 1) use a lambda and wrap it in `NamedObject` or 2) do not wrap the value in `NamedObject`. Wrapping a value that is not callable in `NamedObject` creates the need to use `_obj`, which should be avoided.

This comment applies throughout this PR.

kvignesh1420 replied:
@jsimsa considering the lines mentioned above, if I refactor it to something like this:
```python
def reduce_fn(x, y):
    name, value_fn, expected_structure, expected_types, expected_shapes = y
    return x + combinations.combine(
        value_fn=combinations.NamedObject(
            "value_fn.{}".format(name), value_fn
        ),
        expected_structure=expected_structure,
        expected_types=expected_types,
        expected_shapes=expected_shapes
    )
```
I get the following error:
```console
raise DuplicateTestNameError(test_class_name, new_name, original_name)
absl.testing.parameterized.DuplicateTestNameError: Duplicate parameterized test name in StructureTest: generated test name 
'testFlatStructure_test_mode_eager_tfapiversion_1_expectedshapes__expectedstructure_classdict_expectedtypes_dtypevariant_valuefn_valuefnNested2' (generated from 
'testFlatStructure_test_mode_eager_tfapiversion_1_expectedshapes__expectedstructure_classdict_expectedtypes_dtypevariant_valuefn_valuefnNested2') already exists.
Consider using named_parameters() to give your tests unique names and/or renaming the conflicting test method.
```

kvignesh1420 replied:
Will use lambda's only for this test case as others seem to work fine.

kvignesh1420 replied:
addressed in the commit.

jsimsa reviewed (changes requested):
Thanks. A couple of final comments.

kvignesh1420 committed:
use expected parameter as is in sparse_test

kvignesh1420 committed:
keep lambda usage consistent with NamedObjects

kvignesh1420 commented:
@jsimsa sorry about this long PR. Will try to break up the large change items into multiple PR's next time.
Please let me know how it looks now.

jsimsa reviewed (approved):
Thanks. One minor naming suggestion, otherwise LGTM

kvignesh1420 committed:
rename _evaluate to _checkEqual

kvignesh1420 commented:
@jsimsa renamed the method. Thanks for the review.
